# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1265192305 18000
#      Wed Feb 03 05:18:25 2010 -0500
# Node ID 8cf32587d8f1cc8f0063a4ed93ff7b487fb4077b
# Parent  2fcc927a87573a55d2eb9dcb7729665b101d7165
liboctave/cutils.c (octave_usleep): implement with nanosleep

diff --git a/ChangeLog b/ChangeLog
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,8 +1,14 @@
+2010-02-03  John W. Eaton  <jwe@octave.org>
+
+	* configure.ac: Don't check for poll or usleep.
+	* bootstrap.conf (gnulib_modules): Include nanosleep and sleep
+	in the list.
+
 2010-02-03  John W. Eaton  <jwe@octave.org>
 
 	* configure.ac: Don't check for sigaction, sigpending,
 	sigprocmask, or sigsuspend.
 
 2010-02-03  John W. Eaton  <jwe@octave.org>
 
 	* configure.ac: Don't check for getcwd.
diff --git a/bootstrap.conf b/bootstrap.conf
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -24,23 +24,26 @@ gnulib_modules="
   getcwd
   getopt-gnu
   gettimeofday
   glob
   link
   lstat
   mkdir
   mkfifo
+  sleep
+  nanosleep
   pathmax
   readlink
   rename
   rmdir
   sigaction
   signal
   sigprocmask
+  sleep
   stat
   stdint
   strftime
   strptime
   symlink
   sys_stat
   sys_time
   sys_times
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -1492,17 +1492,17 @@ AC_CHECK_FUNCS(basename canonicalize_fil
   chmod dup2 endgrent endpwent execvp expm1 expm1f fcntl fork \
   getegid geteuid getgid getgrent getgrgid getgrnam getpgrp getpid \
   getppid getpwent getpwuid getuid getwd _kbhit kill \
   lgamma lgammaf lgamma_r lgammaf_r localtime_r log1p log1pf \
   mkstemp pipe poll putenv raise \
   realpath resolvepath rindex roundl select setgrent setlocale \
   setpwent setvbuf siglongjmp \
   snprintf strdup strerror strsignal tempnam tgammaf trunc umask \
-  uname usleep utime vfprintf vsprintf vsnprintf waitpid \
+  uname utime vfprintf vsprintf vsnprintf waitpid \
   _chmod _snprintf x_utime _utime32)
 
 AC_LANG_PUSH(C++)
 AC_CHECK_DECLS([exp2, round, tgamma], [], [], [[#include <cmath>]])
 AC_CHECK_FUNCS([exp2 round tgamma])
 AH_VERBATIM([Z_FUNCS_AND_DECLS], [
 #if defined (__cplusplus)
 extern "C" {
diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,13 @@
+2010-02-03  John W. Eaton  <jwe@octave.org>
+
+	* cutils.c (octave_usleep): Implement using nanosleep.
+	(octave_sleep): Assume sleep exists.
+
 2010-02-03  John W. Eaton  <jwe@octave.org>
 
 	* syscalls.cc (Fcanonicalize_file_name):
 	Call octave_canonicalize_file_name instead of
 	file_ops::canonicalize_file_name.
 
 2010-02-03  John W. Eaton  <jwe@octave.org>
 
diff --git a/src/cutils.c b/src/cutils.c
--- a/src/cutils.c
+++ b/src/cutils.c
@@ -19,96 +19,40 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
-#include <stdarg.h>
 #include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
+#include <time.h>
 
-#if defined (__WIN32__) && ! defined (_POSIX_VERSION)
-
-#include <windows.h>
-
-#else
-
-#include <sys/time.h>
 #include <sys/types.h>
 #include <unistd.h>
 
-#ifdef HAVE_POLL_H
-#include <poll.h>
-#elif HAVE_SYS_POLL_H
-#include <sys/poll.h>
-#endif
-
-#endif
-
 void
 octave_sleep (unsigned int seconds)
 {
-#if defined (__WIN32__) && ! defined (_POSIX_VERSION)
-  Sleep (1000 * seconds);
-#else
   sleep (seconds);
-#endif
 }
 
 void
 octave_usleep (unsigned int useconds)
 {
+  struct timespec delay;
+  struct timespec remaining;
+
   unsigned int sec = useconds / 1000000;
   unsigned int usec = useconds % 1000000;
 
-  if (sec > 0)
-    octave_sleep (sec);
-
-#if defined (__WIN32__) && ! defined (_POSIX_VERSION)
-
-  /* Round to the nearest millisecond, with a minimum of 1 millisecond
-     if usleep was called with a a non-zero value.  */
-
-  if (usec > 500)
-    Sleep ((usec+500)/1000);
-  else if (usec > 0)
-    Sleep (1);
-  else
-    Sleep (0);
-
-#elif defined (HAVE_USLEEP)
-
-  usleep (usec);
-
-#elif defined (HAVE_SELECT)
+  delay.tv_sec = sec;
+  delay.tv_nsec = usec * 1000;
 
-  {
-    struct timeval delay;
-
-    delay.tv_sec = 0;
-    delay.tv_usec = usec;
-
-    select (0, 0, 0, 0, &delay);
-  }
-
-#elif defined (HAVE_POLL)
-
-  {
-    struct pollfd pfd;
-
-    int delay = usec / 1000;
-
-    if (delay > 0)
-      poll (&pfd, 0, delay);
-  }
-
-#endif
+  nanosleep (&delay, &remaining);
 }
 
 int
 octave_raw_vsnprintf (char *buf, size_t n, const char *fmt, va_list args)
 {
   return vsnprintf (buf, n, fmt, args);
 }
